### PR TITLE
fix(android/app): Add storage permission request to "Get Started" menu

### LINKED ISSUE #4120: bug(android/app): Android request for storage permission hidden by "Getting Started"
Starting in Android M, Android requires apps to that use external storage (SD card) to request for storage permission.
The Keyman app uses this for accessing downloaded kmp packages.

@makara noted in user testing the system dialog is hidden by the "Getting Started" menu. On a fresh install of the app, a user could go through the "Getting Started" steps and install a keyboard from keyman.com before seeing and accepting the system dialog.

We should display "Getting Started" after the user deals with the system dialog.

### PR DESCRIPTION
**Pull request by darcywong00**
Fixes #4120 where the Android system dialog for requesting storage permission was getting hidden by the "Get Started" menu.

This adds the check for requesting storage permission to the "Get Started" menu as Step 1.
On Android Lollipop (API 21), this permission isn't needed, so the step defaults to being checked.
On Android Marshmallow onward, this permission needs to be granted in order for Keyman to
* download updates (See #2946)
* install kmp files from the SD card using the device's File browser.

Installing kmp's from the keyboard search will still work when storage permission is denied.

# User Testing

## Android Lollipop (API 21)
1. Do a fresh install of Keyman for Android
2. Verify the "Get Started" menu is displayed like below:
api 21 - get started

3. Verify "Step 1" for "Enable Keyman to access external storage" is already checked
4. Touch "Step 1" and verify nothing happens
5. Touch "Step 2" and proceed with the steps of searching for a keyboard and installing it
6. Verify the keyboard installs

----

### Deny Storage Permission
1. Do a fresh install of Keyman for Android
2. Verify the "Get Started" menu is displayed like below:
api 29 - get started

3. Verify "Step 1" for "Enable Keyman to access external storage" is not checked (shows number "1")
4. Touch "Step 1" and verify request for storage permission is displayed
api 29 - request permission

5. Touch "Deny"
6. Verify "Step 1" for "Enable Keyman to access external storage" is not checked (shows number "1")
7. Touch "Step 1" and verify request for storage permission is displayed again (it has a third option to deny & don't ask again)
api 29 - request permission again
8. Touch "Deny & don't ask again"
9. Verify Step 1 for "Enable Keyman to access external storage" is not checked (shows number "1")
10. Touch "Step 1" and verify nothing happens
11. Close the "Get Started" menu
12. From the "Settings" menu --> Install Keyboard or Dictionary --> Install from keyman.com
13. Verify you can search and install a keyboard
14. From a Chrome browser, download a kmp keyboard package.
15. From the device's File browser, copy that kmp file to the SDCARD --> Download folder
16. From the Keyman "Settings" menu --> Install Keyboard or Dictionary --> Install from local file
17. Browse to the SDCARD --> Download and attempt to install the local kmp file.
18. Verify the keyboard installation succeeds
19. From the device's File browser, browse to SDCARD --> Download --> Click on the kmp file
20. Verify Keyman displays a notification about the installation failing due to storage permission
api 29 - sd kmp fail

## Storage Permission Granted
1.  Do a fresh install of Keyman for Android
2. Verify the "Get Started" menu is displayed like below:
api 29 - get started

3. Verify "Step 1" for "Enable Keyman to access external storage" is not checked (shows number "1")
4. Touch "Step 1" and verify request for storage permission is displayed
api 29 - request permission

5. Touch "Allow"
6. Verify "Step 1" is now checked
api 29 - storage permission granted

7. Touch "Step 2" and proceed with installing a keyboard
8. Verify keyboard installs

### CODE REVIEW

**Comment by darcywong00** [comment]
@MakaraSok can you test this? We'll need to wait for CI to return from maintenance to create a test build.

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/GetStartedActivity.java [hunk 1]
```diff
@@ -149,31 +158,41 @@ public void onWindowFocusChanged(boolean hasFocus) {
...
       String two = String.valueOf(R.drawable.ic_looks_two);
       String three = String.valueOf(R.drawable.ic_looks_three);
+      String four = String.valueOf(R.drawable.ic_looks_four);
       String checkbox_on = String.valueOf(android.R.drawable.checkbox_on_background);
       String info = String.valueOf(R.drawable.ic_info_outline);
 
-      List<Keyboard> kbList = KMManager.getKeyboardsList(this);
-      if (kbList != null && kbList.size() > 1) {
+      boolean storagePermissionGranted = MainActivity.storagePermissionGranted(this);
+      if (storagePermissionGranted) {
```
Comment (darcywong00) [inline android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/GetStartedActivity.java:166]: This new "Step 1" shifts all the previous steps down one row.

**Comment by MakaraSok** [comment]
Tested:

- **Android Lollipop (API 21)** 👍 
- **Marshmallow or greater... Deny Storage Permission**
_On emulator, everything works as expected 👍_
 _On Android 10 running on Samsung Galaxy S9+, at step 20, when opening file browser without the permission granted using the file browser app by Samsung (aka "My File"), this message shows that it seems Keyman is not even installed as it requests the user to find an app on the Play Store to open the file._
image

- **Storage Permission Granted** 👍

**Comment by darcywong00** [comment]
>  using the file browser app by Samsung (aka "My File"), this message shows that it seems Keyman is not even installed as it requests the user to find an app on the Play Store to open the file.

Yes, the Samsung file browser doesn't handle the intents for opening kmp files (Issue #629). That's why we added the "Install from local file" option to the Settings menu.

**Comment by mcdurdin** [comment]
> * Touch "Deny & don't ask again"
> ...
> * Verify you can search and install a keyboard

From the above steps, this makes it seem that storage permission is only required if installing a file from the file system, not when downloading or updating files? If that is the case, then I think we should only prompt for that permission when the user first tries to install a file from the file system, and not at install time.

**Comment by darcywong00** [comment]
> this makes it seem that storage permission is only required if installing a file from the file system, not when downloading or updating files?

If storage permissions are denied, I believe it will generate storage exceptions when attempting to download updates

**Comment by mcdurdin** [comment]
> If storage permissions are denied, I believe it will generate storage exceptions when attempting to download updates

How can we do the initial download then? This seems backwards and I would like to get to the bottom of it -- we don't want to be requesting permissions we don't actually need. And we really shouldn't need external storage perms to download an update if we don't need them to do an install.

**Comment by darcywong00** [comment]
I'll investigate to see if not setting a destination URI will allow the update Download Manager to proceed without exceptions, and then see if I can use 
https://developer.android.com/reference/android/app/DownloadManager#getUriForDownloadedFile(long)
to the the update .kmp file.

**Comment by darcywong00** [comment]
Abandoning this PR in favor of #4170

**Pull request closed without merging**